### PR TITLE
[RFR] Play a single game

### PR DESCRIPTION
**Pull request by Luwangel**
- [x] Create the basic components (buttons, loaders)
- [x] Upgrade the style
- [x] Connect the core to display a shuffled grid

### CODE REVIEW

File: pages/game.js [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+        });
+    };
+
+    render() {
+        const { isLoading, currentGrid, turn, isVictory } = this.state;
+
+        if (isLoading) {
+            return (
+                <div>
+                    <CustomHead />
```
Comment (jdemangeon) [inline pages/game.js]: There's too much repeated code in this page.

I'll extract all common code (CustomHead, Nav, ...) in a dedicated template (thinks about base.html.twig in Symfony for example).

File: pages/index.js [hunk 2]
```diff
@@ -1,10 +1,48 @@
...
+import Nav from '../src/components/nav';
 
 export default class Index extends Component {
+    handleClickSingleGame = () => {};
+
     render() {
         return (
             <div>
-                <h1>Welcome to the 15 puzzle game!</h1>
+                <CustomHead />
```
Comment (jdemangeon) [inline pages/index.js]: Same

File: pages/game.js [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+
+    state = {
+        isLoading: true,
+        currentGrid: null,
+        initialGrid: null,
+        turn: -1,
+        isVictory: false,
+    };
+
+    renderWinnerMessage(isVictory, turn) {
```
Comment (djhi) [inline pages/game.js]: avoid renderXXX methods. Just put the condition directly inside the rendered JSX or extract in another component.

File: src/components/button.js [hunk 4]
```diff
@@ -0,0 +1,25 @@
+import React, { Component } from 'react';
+import Link from 'next/link';
+import PropTypes from 'prop-types';
+
+export default class Button extends Component {
```
Comment (jdemangeon) [inline src/components/button.js]: There's no needs for "Class" Components for simple cases (with no state manipulation) as this button.

You should use Stateless Component in this case (https://hackernoon.com/react-stateless-functional-components-nine-wins-you-might-have-overlooked-997b0d933dbc). It's more concise and clearer.

```
const Button = ({ color, icon, label, path }) => (
    <Link href={{ pathname: path }}>
        <a className={`btn ${color ? color : ''}`}>
            <i className="material-icons left">{icon}</i>
            {label}
        </a>
    </Link>
);

Button.propTypes = {...};

export default Button;
```

File: src/components/activityIndicator.js [hunk 5]
```diff
@@ -0,0 +1,57 @@
+import React, { Component } from 'react';
+
+export default class ActivityIndicator extends Component {
```
Comment (djhi) [inline src/components/activityIndicator.js]: If your component just render stuff without any other logic, make it a functional component

File: src/components/button.js (hunk 4 above)
Comment (djhi) [inline src/components/button.js]: same

File: src/components/grid.js [hunk 6]
```diff
@@ -0,0 +1,49 @@
+import React, { Component } from 'react';
+import PropTypes from 'prop-types';
+
+import Tile from './tile';
+
+export default class Grid extends Component {
```
Comment (djhi) [inline src/components/grid.js]: same

File: src/components/head.js [hunk 7]
```diff
@@ -0,0 +1,25 @@
+import React from 'react';
+import Head from 'next/head';
+
+export default function CustomHead() {
```
Comment (djhi) [inline src/components/head.js]:
```jsx
const CustomHead = () => (JSX);

export default CustomHead;
```

**Review by djhi — CHANGES_REQUESTED** [review]

File: src/components/activityIndicator.js (hunk 5 above)
Comment (djhi) [reply src/components/activityIndicator.js]: Note that this is just a personal preference, not a change request

File: src/components/button.js (hunk 4 above)
Comment (jdemangeon) [reply src/components/button.js]: You're component don't need to re-render each time props changes. If props are the same, there's no reason to re-render. It's called "purity", with "pure" components.

With "class" component, you should extends 
[PureComponent](https://reactjs.org/docs/react-api.html#reactpurecomponent).

With stateless component, you should "pure" HOC from recompose https://github.com/acdlite/recompose/blob/master/docs/API.md#pure

File: src/components/button.js [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+        icon: PropTypes.string.isRequired,
+        label: PropTypes.string.isRequired,
+        path: PropTypes.string.isRequired,
+    };
+
+    render() {
+        const { color, icon, label, path } = this.props;
+
+        return (
+            <Link href={{ pathname: path }}>
```
Comment (jdemangeon) [inline src/components/button.js]: Same for "purity".

File: src/components/button.js (hunk 4 above)
Comment (jdemangeon) [reply src/components/button.js]: I advice you to read the article below. It'll help you to understand performance considerations in React.

https://marmelab.com/blog/2017/02/06/react-is-slow-react-is-fast.html

File: src/components/grid.js [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+    };
+
+    static defaultProps = {
+        readOnly: true,
+    };
+
+    render() {
+        const { grid, readOnly, onPress } = this.props;
+
+        return (
```
Comment (jdemangeon) [inline src/components/grid.js]: Same for "purity"

File: src/components/grid.js [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+            <div className="column">
+                {grid.map((row, rowKey) => {
+                    return (
+                        <div className="row" key={rowKey}>
+                            {row.map(tileValue => {
```
Comment (jdemangeon) [inline src/components/grid.js]: You can use ternary here.

```
tileValue === 0 ? (<div>...</div>) : (<div>...</div>)
```

File: src/components/nav.js [hunk 11]
```diff
@@ -0,0 +1,27 @@
+import React, { Component } from 'react';
+import PropTypes from 'prop-types';
+
+import Link from 'next/link';
+
+export default class Nav extends Component {
```
Comment (jdemangeon) [inline src/components/nav.js]: Same for "purity"

File: __tests__/src/components/button.spec.js [hunk 12]
```diff
@@ -0,0 +1,27 @@
+import React from 'react';
+import { shallow } from 'enzyme';
+import renderer from 'react-test-renderer';
+
+import Button from '../../../src/components/button';
+
+describe('Button', () => {
+    it('it should render the Button without error', () => {
```
Comment (floo51) [inline __tests__/src/components/button.spec.js]: it x2

File: pages/game.js [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+        isLoading: true,
+        currentGrid: null,
+        initialGrid: null,
+        turn: -1,
+        isVictory: false,
+    };
+
+    handleClick = () => {};
+
+    renderWinnerMessage(isVictory, turn) {
```
Comment (floo51) [inline pages/game.js]: Having functions like this is usually a sign that the component should be split. That or use an inline if.

File: pages/game.js [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+                                label="Back to home"
+                                path="/"
+                            />
+                        </Row>
+                    </Section>
+                </Page>
+            );
+        }
+
+        return (
```
Comment (floo51) [inline pages/game.js]: Too many duplicated code, you can merge both returns into one and use JSX:

```jsx
return (
    <div>
          {shouldShowThisParticularThing &&
               <h1> Hello ! </h1>
          }
    </div>
);
```

**Review by floo51 — CHANGES_REQUESTED** [review]

File: __tests__/pages/index.spec.js [hunk 15]
```diff
@@ -4,7 +4,7 @@ import renderer from 'react-test-renderer';
 import Index from '../../pages/index';
 
 describe('Index', () => {
-    it('it should render Index without error', () => {
+    it('should render Index without error', () => {
```
Comment (djhi) [inline __tests__/pages/index.spec.js]: `test`, not `it` if you're using jest

File: __tests__/src/components/activityIndicator.js [hunk 16]
```diff
@@ -4,7 +4,7 @@ import renderer from 'react-test-renderer';
 import ActivityIndicator from '../../../src/components/activityIndicator';
 
 describe('ActivityIndicator', () => {
-    it('it should render ActivityIndicator without error', () => {
+    it('should render ActivityIndicator without error', () => {
```
Comment (djhi) [inline __tests__/src/components/activityIndicator.js]: `test`, not `it` if you're using jest

File: __tests__/src/components/button.spec.js [hunk 17]
```diff
@@ -5,7 +5,7 @@ import renderer from 'react-test-renderer';
 import Button from '../../../src/components/button';
 
 describe('Button', () => {
-    it('it should render the Button without error', () => {
+    it('should render the Button without error', () => {
```
Comment (djhi) [inline __tests__/src/components/button.spec.js]: `test`, not `it` if you're using jest

**Review by djhi — CHANGES_REQUESTED** [review]

**Review by floo51 — APPROVED** [review]

**Review by djhi — APPROVED** [review]

**Pull request merged**